### PR TITLE
JDK-8302335: IGV: Bytecode not showing

### PR DESCRIPTION
**Pull request by tobiasholenstein**
Currently, IGV does not show the bytecode of a graph in the "Bytecode" window although the bytecode is present in the XML file.

# Solution

Following is the `<bytecodes>` block of the XML file
```
<bytecodes>
<![CDATA[
   0 iconst_3
   1 istore_1
   2 bipush 9
   4 istore_2
   5 iload_2
   6 ifle 34
  0   bci: 6    BranchData          taken(0) displacement(112) not taken(0)
   9 iconst_2
  10 istore_3
  11 iload_3
  12 iload_2
  13 if_icmpge 28
  32  bci: 13   BranchData          taken(0) displacement(56)  not taken(0)
  16 iload_1
  17 istore_1
  18 iconst_1
  19 iload_3
  20 irem
  21 istore_1
  22 iinc #3 1
  25 goto 11
  64  bci: 25   JumpData            taken(0) displacement(-32)
  28 iinc #2 -1
  31 goto 5
  88  bci: 31   JumpData            taken(0) displacement(-88)
  34 return
]]>
</bytecodes>
```
The problem was that during paring IGV skipped all lines that started with an empty space: This just skipped all bytecodes. We only want to skip bytecode like `  0   bci: 6    BranchData          taken(0) displacement(112) not taken(0)` . Meaning: whitespace - number - exactly two white spaces - and then arbitrary text.

bytecode

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8302335](https://bugs.openjdk.org/browse/JDK-8302335): IGV: Bytecode not showing

### Reviewers
 * [Roberto Castañeda Lozano](https://openjdk.org/census#rcastanedalo) (@robcasloz - **Reviewer**)
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**)
 * [Xin Liu](https://openjdk.org/census#xliu) (@navyxliu - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/12537/head:pull/12537` \
`$ git checkout pull/12537`

Update a local copy of the PR: \
`$ git checkout pull/12537` \
`$ git pull https://git.openjdk.org/jdk pull/12537/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12537`

View PR using the GUI difftool: \
`$ git pr show -t 12537`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12537.diff">https://git.openjdk.org/jdk/pull/12537.diff</a>

</details>

### CODE REVIEW

**Review by robcasloz — CHANGES_REQUESTED** [review]
Good catch! The issue with the more liberal matching of bytecodes proposed in this PR is that bytecode attributes are now also matched and treated as bytecodes, see e.g. the second last line in this screenshot:

bytecodes

Somewhere between JDK 20 b17 and JDK 20 b25 (most likely with the integration of [JDK-8292699](https://bugs.openjdk.org/browse/JDK-8292699)), the indentation in bytecode dumps produced by `BytecodeTracer::print_method_codes()` changed. The change introduced indentation for the bytecodes but did not update the indentation of their attributes, which I guess was unintended. Here is an example:

Before:

```
0 nofast_aload_0
1 invokevirtual 478 <java/lang/String.isLatin1()Z> 
  0   bci: 1    VirtualCallData     count(5119) nonprofiled_count(0) entries(0)

```

After:

```
   0 nofast_aload_0
   1 invokevirtual 478 <java/lang/String.isLatin1()Z> 
  0   bci: 1    VirtualCallData     count(5119) nonprofiled_count(0) entries(0)

```

I am not sure about what is the best solution to this, but ideally we would want to support the old and current indentation, and probably also fix and support the proper indentation of bytecode attributes in `BytecodeTracer::print_method_codes()`.

A possible solution from IGV side is to rule out attribute lines, e.g. likes starting with  a number then the `bci` keyword such as 
`  56  bci: 4    BranchData          taken(0) displacement(48)`
or lines starting with non-digits such as
`                                    not taken(5119)`
And so on. Another possible solution is to actually parse these attributes and display them somehow (e.g. as bytecode sub-items), since they actually convey useful information. What do you think?

**Comment by tobiasholenstein** [comment]
> Good catch! The issue with the more liberal matching of bytecodes proposed in this PR is that bytecode attributes are now also matched and treated as bytecodes, see e.g. the second last line in this screenshot:
> 
> bytecodes
> 
> Somewhere between JDK 20 b17 and JDK 20 b25 (most likely with the integration of [JDK-8292699](https://bugs.openjdk.org/browse/JDK-8292699)), the indentation in bytecode dumps produced by `BytecodeTracer::print_method_codes()` changed. The change introduced indentation for the bytecodes but did not update the indentation of their attributes, which I guess was unintended. Here is an example:
> 
> Before:
> 
> ```
> 0 nofast_aload_0
> 1 invokevirtual 478 <java/lang/String.isLatin1()Z> 
>   0   bci: 1    VirtualCallData     count(5119) nonprofiled_count(0) entries(0)
> ```
> 
> After:
> 
> ```
>    0 nofast_aload_0
>    1 invokevirtual 478 <java/lang/String.isLatin1()Z> 
>   0   bci: 1    VirtualCallData     count(5119) nonprofiled_count(0) entries(0)
> ```
> 
> I am not sure about what is the best solution to this, but ideally we would want to support the old and current indentation, and probably also fix and support the proper indentation of bytecode attributes in `BytecodeTracer::print_method_codes()`.
> 
> A possible solution from IGV side is to rule out attribute lines, e.g. likes starting with a number then the `bci` keyword such as ` 56 bci: 4 BranchData taken(0) displacement(48)` or lines starting with non-digits such as ` not taken(5119)` And so on. Another possible solution is to actually parse these attributes and display them somehow (e.g. as bytecode sub-items), since they actually convey useful information. What do you think?

You are right. I changed it such that we now support the old and new format. Now we just skip lines with the `bci:` keyword.
Regarding parsing and displaying bci information - I think this should be done in a different RFE

**Comment by navyxliu** [comment]
>   0   bci: 1    VirtualCallData     count(5119) nonprofiled_count(0) entries(0)

Those are methoddata.  Should we add a leading character such as '#' to distinct them from real bytecode?  I believe that would simplify matching logic in IGV.

**Comment by tobiasholenstein** [comment]
> methoddata

This would require a VM change for a printing flag. E.g. `-XX:+CIPrintMethodCodes` uses that code as well - I think we should leave it as it is because the main use is intended to be human-readable and it could break parsing of other uses (that we are not aware of)

**Comment by tobiasholenstein** [comment]
> > 0   bci: 1    VirtualCallData     count(5119) nonprofiled_count(0) entries(0)
> 
> Those are methoddata. Should we add a leading character such as '#' to distinct them from real bytecode? I believe that would simplify matching logic in IGV.

This would require a VM change for a printing flag. E.g. `-XX:+CIPrintMethodCodes` uses that code as well - I think we should leave it as it is because the main use is intended to be human-readable and it could break parsing of other uses (that we are not aware of)

File: src/utils/IdealGraphVisualizer/Data/src/main/java/com/sun/hotspot/igv/data/InputMethod.java [hunk 1]
```diff
@@ -111,10 +111,11 @@ public String getShortName() {
 
     public void setBytecodes(String text) {
         Pattern instruction = Pattern.compile("\\s*(\\d+)\\s*:?\\s*(\\w+)\\s*(.*)(?://(.*))?");
+        Pattern isInfo = Pattern.compile("(\\s*)(\\d+)(\\s+)(bci:)(.+)");
         String[] strings = text.split("\n");
         int oldBci = -1;
         for (String s : strings) {
-            if (s.startsWith(" ")) {
+            if (isInfo.matcher(s).matches()) {
                 // indented lines are extra textual information
```
Comment (robcasloz) [inline src/utils/IdealGraphVisualizer/Data/src/main/java/com/sun/hotspot/igv/data/InputMethod.java:119]: Maybe update this comment?

**Review by robcasloz — APPROVED** [review]
Looks good otherwise!

**Comment by TobiHartmann** [comment]
The method data still shows up for me:

Screenshot from 2023-02-16 13-39-18

**Comment by tobiasholenstein** [comment]
> The method data still shows up for me:
> 
> Screenshot from 2023-02-16 13-39-18



> The method data still shows up for me:
> 
> Screenshot from 2023-02-16 13-39-18

Thanks for catching that! 
This is actually a bug in printing the bytecodes on the VM side: [JDK-8302656](https://bugs.openjdk.org/browse/JDK-8302656)
I adjusted the PR to work even in the presents of the bug (and still work when it is fixed)

**Review by robcasloz — APPROVED** [review]

**Review by navyxliu — APPROVED** [review]
LGTM. I am not a reviewer.

**Comment by tobiasholenstein** [comment]
thanks @robcasloz , @TobiHartmann and @navyxliu for the reviews! 

/integrate

**Pull request closed without merging**